File: dsa self paced problems/Recursion/subset.java
class Main
{
  public static void main(String args[])
  {
    StringBuffer str=new StringBuffer("abc");
    StringBuffer a=new StringBuffer("");
    Main.subset(str,0,a);
  }
  static void subset(StringBuffer str, int index, StringBuffer string) 
  {
    if(index==str.length())
    {
        System.out.println(string+", ");
        return;
    }
      subset(str, index + 1, string);
      //string.append(str.charAt(index));
      subset(str, index + 1, string.append(str.charAt(index)));
      // string.replace(index, index+1, str.sub)
    return;
  }
}

File: dsa self paced problems/Recursion/subset.cpp
#include<iostream>
#include<string.h>
using namespace std;
void subset(string str, int index, string current)
{
    if(index==str.length())
    {
        cout<<current<<", ";
        return;
    }
    subset(str,index+1,current); // if index become same as of the length of the string, the recursion will fall into an infinity loop and never stop, never each the condition of printing statement.
    subset(str,index+1,current+str[index]);
}
int main()
{
  string str="abc";
  subset(str,0,"");
  return 0;
}